README.md
# Captain's Log
*"Bring explicit intentionality into your life."* \
Started: Thu - 3/18/21

npm init \
npm install eslint --save-dev \
npx eslint --init

### Set up an `upstream` remote.
git remote add upstream https://github.com/ORIGINAL_OWNER/ORIGINAL_REPOSITORY.git

### Reset my fork to be up-to-date with original repo.
git reset --hard upstream/master

### To hard-pull from HEAD and overwrite all local files.
git fetch \
git reset --hard origin/main \
git reset --hard upstream/master

### Git commands
git log origin/main..HEAD \
git remote -v

### Firebase Commands
firebase emulators:start --only "auth,firestore" --import=data\multi-users --export-on-exit=data\saved_data \
firebase emulators:start --only hosting \
firebase deploy --only hosting \
firebase deploy --only functions \
firebase emulators:export data\multi-users

- https://stackoverflow.com/questions/1125968/how-do-i-force-git-pull-to-overwrite-local-files
- https://stackoverflow.com/questions/39632667/how-do-i-kill-the-process-currently-using-a-port-on-localhost-in-windows
- https://stackoverflow.com/questions/65586212/how-to-shutdown-firebase-emulator-properly-on-windows-10

```shell
$> netstat -ano | findstr :8080
$> taskkill /PID <PID> /F
$> dir env:
```

### Misc Links
- https://mariusschulz.com/blog/declaring-global-variables-in-typescript
- https://stackoverflow.com/questions/52100103/getting-all-documents-from-one-collection-in-firestore
- https://medium.com/firebase-developers/the-secrets-of-firestore-fieldvalue-servertimestamp-revealed-29dd7a38a82b
- https://gitmoji.dev/

---

## Project Motivation
Several motivations inspired and drove this project.

First: I've always been a big fan of the "self-quantification movement." I own a Fitbit and enjoy seeing granular metrics on everything I do: Sleep, exercise, etc. But Fitbit alone was insufficient. I wanted to know how much time I spent eating, brushing my teeth, running errands, etc.

Second: I wanted a good project with which to learn React with hooks and what webdev looks like nowadays in 2021. Webpack, ESLint, TypeScript, Firebase preview channels, the whole nine yards. My previous knowledge on the webdev front was horribly outdated. Also: Tests, design patterns, Git. It was high time to get caught up.

Finally, realistically though, what ***really*** drove this project: My wife, honest to God, bugs me probably once a week that "I don't spend enough time with her." And that: "I love my computer and spend more time with it than I love her."  Utter baloney.  With this app, I now finally have hard, cold data to show The Wife that, "See? We *did* visit the park four times last week!" and "I spent an aggregate total of __ hours __ minutes with you last month watching Netflix, going out for walks, and eating."

Data and technology to the rescue! 😀

changelog.md
# Changelog

## [Sun - 4/25/21 | Sprint Grape 🍇](https://github.com/r002/captains-log/pull/45)
- 🔖 0.6.17: 🐞 Fix another `Line Break Separator` GUI bug
- 🔖 0.6.16: ✨ Add `footer` that shows the build's `version number` & `build date`
- 🐞 Fix `Line Break Separator` GUI bug
- ✨ Show an `🍿` icon on card if the card's been updated in the past `1 hour`
- ✨ Add `CodeNewbie` link in `Topbar` navigation
- ✨ Add `Topbar` navigation
- 🐞 Fix `Streak Counter` bug
- 💄♻️ Make minor cosmetic fixes and code optimizations
- ✨ Add giant `Countdown Clock` and impl `Streak Counter` + minor cosmetic fixes
- ♻️ Fix the ad-hoc code that adjusts Anita's card dates for #16, #19, #22
- ✨ Add `progress column` for Matt and study group member photos
- ♻️ Refactor `Map<string, Map<string, Card>>` into a proper `UserProgress.ts` model
- ✨ Add `StudyGroup` entry point and POC
- ✨ Add feature to make `Sidebar.tsx` menu items collapsible
- ♻️ Move static `.html` entry points into `/public` dir
- ✨ Impl populating the sidebar ('GTx ISYE6501' menu items) from Firestore
- ✨ Add '/scripts' and 'firebase-admin' in order to easily populate Firestore
- ♻️ Rename `Secure.tsx` to `FileViewer.tsx`. Properly implement `FileViewer.tsx`
- 💄 Beautify the 'Secure Asset' page to show a nice message if user is unauthorized to view asset
- ✨ Add Cloud Function to deliver secure assets to authorized users
- ✨ Add 'Secure' page-- add ability to read protected assets from Firebase Storage
- ♻️ Change how `flashAlert` is implemented (refactor this one day to make correct 🤦‍♂️)
- 0.6.0: 🍇 Begin Sprint Grape
- TODO: Setup a listener to a firestore collection: `realtimeNotifications`
- TODO: Set `deadline` from admin console which pushes to all client endpoints in real-time
- TODO: After deadline passes, show results of the voting
- TODO: Implement Comments

## [Sat - 4/10/21 | Sprint Fig 🥝](https://github.com/r002/captains-log/pull/41)
- ✨ Begin Sprint Fig
- ✨ Add basic Admin tools
- 🗃 Add Firestore mock data for 'passages'
- ✨ Add 'Candidates' and 'Voting' buttons
- ✨ Begin implmenting 'voting record' functionality
- ✨ Done: After you vote, disable the voting buttons (show what you voted for)
- 🛠 Fix webpack's HMR
- ✨ Add a 'Write' page where users can submit passage candidates (WIP)
- ♻️ Modify 'App.tsx' to bounce all unautheticated users
- ✨ Add sidebar to 'Write' page (WIP)
- ♻️ Move 'changelog' into its own 'changelog.md' file
- ✨ Add overall collapsible sidebar to all pages (WIP)
- 💄 Beautify the sidebar
- ✨ Impl 'deleteVote' firestore api call (WIP)
- ♻️ Change how votes are saved/read to/from firestore
- ✨ Add 'Results' page (WIP)
- ✨ Change the url in the address bar upon page navigation
- ✨ In sidebar: Highlight the page the user is currently looking at
- ♻️ Move 'page' widgets into their own 'pages' dir
- ✨ In sidebar: Add links to 'Admin Console' & 'Write' pages
- ✨ Sidebar: Add ability to navigate to external links
- ✨ Sidebar: Add external link to 'Changelog'
- ✨ Sidebar: Don't render when user is logged out
- ✨ 'Vote' page: Display a giant countdown clock until next coronation

## [Fri - 4/9/21 | Sprint Elderberry 🍒](https://github.com/r002/captains-log/pull/40)
- ✨ Begin adding Wattpad POC page. Add new 'Wattpad' entrypoint to webpack
- 🛂 Add Clark Kent dummy user, update Firestore rules, handle unauthorized users
- ♻️ Move FlashAlert inside NavBar widget
- ✨ Add StoryBoard widget
- 🗃 Update Firestore mock data


## [Thu - 4/1/21 | Sprint Durian 🍈](https://github.com/r002/captains-log/pull/33)
- ♻️ Refactor widgets into 'LogTypes' & 'Inputs' for better organization</li>
- 🛠 Fix package-lock.json</li>
- 🗃 Update 'data/multi-users' with YoutubeLogs dummy data</li>
- ♻️ #27: Begin refactoring global event handlers for 'log actions' into DataContext</li>
- 🔨 Add npm dev script shortcut to start auth/firestore emulators</li>
- ♻️ #27: Finish refactoring global event handlers for 'log actions' into DataContext</li>
- ✨ #31: Add 'Details Pane'. Shows Youttube videos for YT log entries</li>
- ♻️ Change it so 'Welcome Log' is always shown if no user is logged in</li>
- ✨ #31: Improve the 'Details Pane'-- show different pane for ActivityLogs</li>
- ✨ #38: Add up/down keyboard arrow support to navigate logs</li>
- ✨ Add support for parsing 'https://youtu.be/...' vids</li>

## [Mon - 3/29/21 | Sprint Apple 🍎](https://github.com/r002/captains-log/pull/24)
- 🐞 Fix #23: 'FlashAlert' bug. Reset every time the theme changes</li>
- ♻️ Refactor 'ThemeContext' to expose a 'toggler' hook</li>
- ♻️ Refactor the buttons in Navbar to use types</li>
- ✨ #4: Add 'watch' command to add clickable Youtube links in logs</li>
- ♻️ Refactor public config into its own file</li>
- 🐞 Fix IooB bug when processing ActivityLogs to compute durations</li>
- 🐞 Fix bug: Allow dt of YoutubeLogs to be edited</li>
- ♻️ Refactor the way logs are read from db & updated. Use 'Object.assign(..)'</li>
- ♻️🤡 Refactor code in 'Navbar'. Add 'Export Logs' dummy button</li>
- ♻️🤡 Rename all event handlers in 'Navbar'. Add 'Export Logs' dummy function</li>
- ✨ #29: Impl log exporter - exports the most recent 1000 logs as '.json' file</li>
- ♻️ Refactor the way YoutubeLogs are stored in db. Store 'vid' instead of 'url'</li>

## [Fri - 3/26/21 | PR #18: Add dt/activity edit/delete functionality in ActivityLog](https://github.com/r002/captains-log/pull/18)
- 🧹 Delete '\data\saved_data' from repo</li>
- 🙈 Update .gitignore to exclude local Firestore data that's saved on exit (`/data/saved_data/`)</li>
- ♻️ Refactor code to generate unique Firestore ids on the clientside before writing to Firestore</li>
- 🤡 Add dummy GUI buttons in ActivityLog widget</li>
- ✨ Impl 'delete button' (❌) in ActivityLog widget</li>
- ♻️ Refactor Firestore api calls into /FirestoreApi.ts</li>
- ♻️ Refactor Ticker into its own widget</li>
- ✨ Impl 'double-click-to-edit/update-activity' for ActivityLog</li>
- ♻️ Combine 'updateLog' into 'writeLog' in /FirestoreApi.ts</li>

scripts/populateMenu.js
const admin = require('firebase-admin')
const serviceAccount = require('../../service-account.json')
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount)
})
const db = admin.firestore()

// db.collection('menus').doc('gtx_isye6501').get().then(items => {
//   console.log('>> items.data:', items.data())
// })

const data = {
  menu: [
    {
      asset: 'GTx__ISYE6501/Timeline1-11Summer__1_.pdf',
      order: 0,
      title: '📅 ISyE 6501x - Timeline (Summer 2020)'
    },
    {
      asset: 'GTx__ISYE6501/Syllabus-combined-Summer2020__3_.pdf',
      order: 1,
      title: '💡 ISyE 6501x - Syllabus (Summer 2020)'
    },
    {
      asset: 'GTx__ISYE6501/week_1_hw-summer.pdf',
      order: 2,
      title: '📄 HW #1: Classification & Validation'
    },
    {
      asset: 'secure/gtx/isye6501/GTx-ISYE6501__hw01-final.pdf',
      order: 3,
      title: '🔒 HW #1: Submission (May 21, 2020)'
    },
    {
      asset: 'secure/gtx/isye6501/week_1_solutions-summer.pdf',
      order: 4,
      title: '🔒 HW #1: Official Solutions'
    },
    {
      asset: 'GTx__ISYE6501/week_2_hw-summer.pdf',
      order: 5,
      title: '📄 HW #2: Clustering, Data Prep, Δ Detection'
    },
    {
      asset: 'secure/gtx/isye6501/week02c_final.pdf',
      order: 6,
      title: '🔒 HW #2: Submission (May 28, 2020)'
    },
    {
      asset: 'secure/gtx/isye6501/week_2_solutions-summer.pdf',
      order: 7,
      title: '🔒 HW #2: Official Solutions'
    },
    {
      asset: 'GTx__ISYE6501/week_3_hw-summer.pdf',
      order: 8,
      title: '📄 HW #3: Time Series + Basic Regression'
    },
    {
      asset: 'secure/gtx/isye6501/week03_subB.pdf',
      order: 9,
      title: '🔒 HW #3: Submission (June 4, 2020)'
    },
    {
      asset: 'secure/gtx/isye6501/week_3_solutions-summer.pdf',
      order: 10,
      title: '🔒 HW #3: Official Solutions'
    },
    {
      asset: 'GTx__ISYE6501/week_4_hw-summer.pdf',
      order: 11,
      title: '📄 HW #4: Adv Data Prep + Adv Regression'
    },
    {
      asset: 'secure/gtx/isye6501/week04_subA.pdf',
      order: 12,
      title: '🔒 HW #4: Submission (June 11, 2020)'
    },
    {
      asset: 'secure/gtx/isye6501/week_4_solutions-summer.pdf',
      order: 13,
      title: '🔒 HW #4: Official Solutions'
    },
    {
      asset: 'GTx__ISYE6501/week_5_hw-summer.pdf',
      order: 14,
      title: '📄 HW #5: Var Sel, Exp Design, Prob Models'
    },
    {
      asset: 'secure/gtx/isye6501/week05_subA.pdf',
      order: 15,
      title: '🔒 HW #5: Submission (June 18, 2020)'
    },
    {
      asset: 'secure/gtx/isye6501/week_5_solutions-summer.pdf',
      order: 16,
      title: '🔒 HW #5: Official Solutions'
    },
    {
      asset: 'GTx__ISYE6501/week_6_hw-summer.pdf',
      order: 17,
      title: '📄 HW #6: Prob Mods, Data Imp, Optimization'
    },
    {
      asset: 'secure/gtx/isye6501/week06_subC.pdf',
      order: 18,
      title: "🔒 HW #6: Robert's Airport PySim (Jun 25, 20)"
    },
    {
      asset: 'secure/gtx/isye6501/week_6_solutions-summer.pdf',
      order: 19,
      title: '🔒 HW #6: Official Solutions'
    },
    {
      asset: 'GTx__ISYE6501/week_7_hw-summer.pdf',
      order: 20,
      title: '📄 HW #7: More Optimization + Adv Models'
    },
    {
      asset: 'secure/gtx/isye6501/week07_subA.pdf',
      order: 21,
      title: '🔒 HW #7: Army Diet Problem (July 1, 2020)'
    },
    {
      asset: 'secure/gtx/isye6501/week_7_solutions-summer.pdf',
      order: 22,
      title: '🔒 HW #7: Official Solutions'
    },
    {
      asset: 'GTx__ISYE6501/week_8_hw-summer.pdf',
      order: 23,
      title: '📄 HW #8: Power Company Case Study'
    },
    {
      asset: 'secure/gtx/isye6501/week08_subA.pdf',
      order: 24,
      title: '🔒 HW #8: Robert PowerCo (July 9, 2020)'
    },
    {
      asset: 'GTx__ISYE6501/week_9_hw-summer.pdf',
      order: 25,
      title: '📄 HW #9: Retailer Case Study'
    },
    {
      asset: 'GTx__ISYE6501/week_10_hw-summer.pdf',
      order: 26,
      title: '📄 HW #10: Monetization Case Study'
    },
    {
      asset: 'GTx__ISYE6501/ISyE_6501_Project.pdf',
      order: 27,
      title: '📄 Final Course Project (July 16, 2020)'
    },
    {
      asset: 'secure/gtx/isye6501/project_subB.pdf',
      order: 28,
      title: '🔒 Final Project: Eprimo GmbH’s Chat Bot'
    },
    {
      asset: 'https://credentials.edx.org/records/programs/shared/5e01b7235a294e56bfef1c2f0f3bd3b4/',
      order: 29,
      title: "🅰️ My Final Grade-- I got an 'A'! (Aug 9, 2020)"
    },
    {
      asset: 'https://courses.edx.org/certificates/0fc4e6b0ad184913a4a31c27d84b7a19',
      order: 30,
      title: '🏆 My Course Certificate (Aug 10, 2020)'
    }
  ]
}

// Add a new document in collection "cities" with ID 'LA'
db.collection('menus').doc('gtx_isye6501').set(data).then(() => {
  console.log('>> Script Finished! | this is populate menu!!!')
})
